### PR TITLE
Clean-up most dead asssignments

### PR DESCRIPTION
**Pull request by ibuclaw**
In a red mood at the moment, so thought I might remove some small bits and pieces found by `-Wunused`, probably are beneficial to DDMD as unlike C++ compilers is not smart enough to remove these dead pieces of code.

### CODE REVIEW

**Comment by 9rnsr** [comment]
Auto-merge toggled on

File: src/template.c [hunk 1]
```diff
@@ -8108,7 +8108,7 @@ void TemplateMixin::semantic(Scope *sc)
     printf("\tdo semantic() on template instance members '%s'\n", toChars());
 #endif
     Scope *sc2 = argscope->push(this);
-    size_t deferred_dim = Module::deferred.dim;
+    //size_t deferred_dim = Module::deferred.dim;
```
Comment (yebblies) [inline src/template.c]: Mistake?

Comment (ibuclaw) [inline src/template.c]: It's kept for documentation purposes - see what is written 10 lines or so down.  (https://github.com/D-Programming-Language/dmd/commit/266b842acba9da7e63c8c59ef798cc4d011caf94)

**Comment by WalterBright** [comment]
Dead assignments are removed by the optimizer, not the front end, even in D compilers, and even in DMD.

**Comment by ibuclaw** [comment]
> Dead assignments are removed by the optimizer, not the front end, even in D compilers, and even in DMD.

Maybe I've missed something, as I've never seen DMD remove dead assignments - even when RHS has no side effects - unless I give the function being called every possible hint to let the compiler know that it truly is side effect free (``pure nothrow @safe``).  GDC (and I guess that includes LDC too) can simply infer.

**Comment by WalterBright** [comment]
> I've never seen DMD remove dead assignments
```
int foo(int x) {
   int y = x + 3;
   return x * x;
}
```
```
dmd foo -c -O
obj2asm foo.obj
```
_D3foo3fooFiZi  comdat
                push    EAX
                imul    EAX,EAX
                add     ESP,4
                ret
```

**Comment by yebblies** [comment]
I would guess Iain was talking about assignments to globals, unless I'm mistaken and dmd can remove those too.

**Comment by ibuclaw** [comment]
> I would guess Iain was talking about assignments to globals

Assignments to globals should *never* be optimised away ???

I'm referring to where RHS is a function call that has no visible side effect in the body, but is not marked as ``pure nothrow`` in code.  There's no example of that in this PR, but we can safely say that dead code such as  ``Type *tb2next = tb2->nextOf();`` we can do without.  I'm not sure off the top of my head whether or not g++ would execute ``tb2->nextOf()`` in this case, but I'm sure that DMD sure will.

**Pull request merged**